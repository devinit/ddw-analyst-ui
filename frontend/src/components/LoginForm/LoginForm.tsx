import classNames from 'classnames';
import { Formik, FormikActions, FormikProps } from 'formik';
import * as React from 'react';
import { Alert, Button, Card, Form, InputGroup } from 'react-bootstrap';
import styled from 'styled-components';
import * as Yup from 'yup';
import SocialLine from '@bit/devinit.analyst-ui.components.social-line';

interface LoginFormState {
  credentials: Credentials;
  alerts: Partial<Credentials>;
}

export interface Credentials {
  username: string;
  password: string;
}

interface LoginFormProps {
  showForm?: boolean;
  alert?: string;
  onSuccess: (values: Credentials, formikActions: FormikActions<Credentials>) => void;
}

const StyledFormFeedback = styled(Form.Control.Feedback)`margin-left: 55px`;
const StyledAlert = styled(Alert)`margin-left: 15px`;

export class LoginForm extends React.Component<LoginFormProps, LoginFormState> {
  static defaultProps: Partial<LoginFormProps> = { showForm: true };
  state: LoginFormState = {
    credentials: { username: '', password: '' },
    alerts: {}
  };
  private schema = Yup.object().shape({
    username: Yup.string().required('Username is required!'),
    password: Yup.string().required('Password is required!')
  });

  render() {
    const initialValues: Credentials = { username: '', password: '' };
    return (
      <Formik validationSchema={ this.schema } initialValues={ initialValues } onSubmit={ this.props.onSuccess }>
        {
          ({ errors, handleChange, handleSubmit, touched }: FormikProps<Credentials>) => (
            <Form className="form" noValidate onSubmit={ handleSubmit } data-testid="login-form">
              <Card className={ classNames('card-login', { 'card-hidden': !this.props.showForm }) }>
                <Card.Header className="card-header-rose text-center">
                  <Card.Title>Login</Card.Title>
                  <SocialLine/>
                </Card.Header>
                <Card.Body>
                  <p className="card-description text-center d-none">Or Go With</p>
                  <StyledAlert variant="danger" hidden={ !this.props.alert }>
                    { this.props.alert }
                  </StyledAlert>
                  <span className="bmd-form-group">
                    <InputGroup>
                      <InputGroup.Prepend>
                        <InputGroup.Text><i className="material-icons">email</i></InputGroup.Text>
                      </InputGroup.Prepend>
                      <Form.Control
                        required
                        name="username"
                        type="email"
                        placeholder="Email..."
                        onChange={ handleChange }
                        isInvalid={ !!errors.username }
                      />
                      <StyledFormFeedback type="invalid">
                        { touched.username && errors.username ? errors.username : null }
                      </StyledFormFeedback>
                    </InputGroup>
                  </span>
                  <span className="bmd-form-group">
                    <InputGroup>
                      <InputGroup.Prepend>
                        <InputGroup.Text><i className="material-icons">lock_outline</i></InputGroup.Text>
                      </InputGroup.Prepend>
                      <Form.Control
                        name="password"
                        type="password"
                        placeholder="Password..."
                        onChange={ handleChange }
                        isInvalid={ !!errors.password }
                      />
                      <StyledFormFeedback type="invalid">
                        { touched.password && errors.password ? errors.password : null }
                      </StyledFormFeedback>
                    </InputGroup>
                  </span>
                </Card.Body>
                <Card.Footer className="justify-content-center">
                  <Button variant="link" size="lg" className="btn-rose" type="submit">Login</Button>
                </Card.Footer>
              </Card>
            </Form>
          )
        }
      </Formik>
    );
  }
}
