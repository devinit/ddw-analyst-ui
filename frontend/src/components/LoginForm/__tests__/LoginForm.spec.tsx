import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { LoginForm } from '../LoginForm';

test('renders the correctly with the default props', () => {
  const renderer = TestRenderer
    .create(<LoginForm/>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders a card with the class card-hidden when showForm is set to false', () => {
  const renderer = TestRenderer
    .create(<LoginForm showForm={ false }/>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders an alert', () => {
  const renderer = TestRenderer
    .create(<LoginForm alert="This is an alert message!"/>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});
