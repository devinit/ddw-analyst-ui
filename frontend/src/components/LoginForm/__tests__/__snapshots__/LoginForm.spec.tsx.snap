// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a card with the class card-hidden when showForm is set to false 1`] = `
<form
  className="form"
  data-testid="login-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="card-login card-hidden card"
  >
    <div
      className="card-header-rose text-center card-header"
    >
      <div
        className="card-title h5"
      >
        Login
      </div>
      <div
        className="social-line"
      />
    </div>
    <div
      className="card-body"
    >
      <p
        className="card-description text-center d-none"
      >
        Or Go With
      </p>
      <div
        className="fade sc-AxirZ fZRTbk alert alert-danger show"
        hidden={true}
        role="alert"
      />
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                email
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            placeholder="Email..."
            required={true}
            type="email"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                lock_outline
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Password..."
            type="password"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
    </div>
    <div
      className="justify-content-center card-footer"
    >
      <button
        className="btn-rose btn btn-link btn-lg"
        disabled={false}
        type="submit"
      >
        Login
      </button>
    </div>
  </div>
</form>
`;

exports[`renders an alert 1`] = `
<form
  className="form"
  data-testid="login-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="card-login card"
  >
    <div
      className="card-header-rose text-center card-header"
    >
      <div
        className="card-title h5"
      >
        Login
      </div>
      <div
        className="social-line"
      />
    </div>
    <div
      className="card-body"
    >
      <p
        className="card-description text-center d-none"
      >
        Or Go With
      </p>
      <div
        className="fade sc-AxirZ fZRTbk alert alert-danger show"
        hidden={false}
        role="alert"
      >
        This is an alert message!
      </div>
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                email
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            placeholder="Email..."
            required={true}
            type="email"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                lock_outline
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Password..."
            type="password"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
    </div>
    <div
      className="justify-content-center card-footer"
    >
      <button
        className="btn-rose btn btn-link btn-lg"
        disabled={false}
        type="submit"
      >
        Login
      </button>
    </div>
  </div>
</form>
`;

exports[`renders correctly with the default props 1`] = `
<form
  className="form"
  data-testid="login-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="card-login card"
  >
    <div
      className="card-header-rose text-center card-header"
    >
      <div
        className="card-title h5"
      >
        Login
      </div>
      <div
        className="social-line"
      />
    </div>
    <div
      className="card-body"
    >
      <p
        className="card-description text-center d-none"
      >
        Or Go With
      </p>
      <div
        className="fade sc-AxirZ fZRTbk alert alert-danger show"
        hidden={true}
        role="alert"
      />
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                email
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            placeholder="Email..."
            required={true}
            type="email"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
      <span
        className="bmd-form-group"
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-prepend"
          >
            <span
              className="input-group-text"
            >
              <i
                className="material-icons"
              >
                lock_outline
              </i>
            </span>
          </div>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Password..."
            type="password"
          />
          <div
            className="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
    </div>
    <div
      className="justify-content-center card-footer"
    >
      <button
        className="btn-rose btn btn-link btn-lg"
        disabled={false}
        type="submit"
      >
        Login
      </button>
    </div>
  </div>
</form>
`;
